src/Common/Core/Impl/Threading/AsyncCountdownEvent.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core.Disposables;
+
+namespace Microsoft.Common.Core.Threading {
+    public class AsyncCountdownEvent {
+        private readonly AsyncManualResetEvent _mre = new AsyncManualResetEvent();
+        private int _count;
+
+        public AsyncCountdownEvent(int initialCount) {
+            if (initialCount < 0) {
+                throw new ArgumentOutOfRangeException(nameof(initialCount));
+            }
+
+            _count = initialCount;
+            if (initialCount == 0) {
+                _mre.Set();
+            }
+        }
+
+        public Task WaitAsync() => _mre.WaitAsync();
+
+        public void Signal() {
+            if (_count <= 0) {
+                throw new InvalidOperationException();
+            }
+
+            var count = Interlocked.Decrement(ref _count);
+            if (count < 0) {
```

Comment:
Not sure that I understand the question, but there two checks are actually a protection. If 2 signals are called concurrently when `_count == 1`, after decrement one of them will fail on `count < 0` check

---
In reply to: [78201558](https://github.com/Microsoft/RTVS/pull/2268#discussion_r78201558) [](ancestors = 78201558)